h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java

```diff
@@ -102,30 +103,31 @@ else if (f.vec(i-catCols).isInt())
     }
 
     /**
-     * Counts a total sum of chunks inside a vector. Only chunks present in chunkIds are considered.
+     * Counts a total sum of chunks inside a vector. Only chunks present in chunkIds are counted.
+     * If a weights vector is provided, only rows with non-zero weights are counted.
      *
-     * @param chunkIds Chunk identifier of a vector
+     * @param chunkIds Chunk ids to consider during the calculation. Chunks IDs not listed are not included.
      * @param vec      Vector containing given chunk identifiers
-     * @param weightsVector Vector with row weights, possibly null
+     * @param weightsVector Vector with row weights, possibly an empty optional
+     * @param chunkLengths Array of integers where the lengths of the individual chunks will be added. Initialization to an array of 0's is expected.
      * @return A sum of chunk lengths. Possibly zero, if there are no chunks or the chunks are empty.
      */
-    public static long sumChunksLength(int[] chunkIds, Vec vec, Vec weightsVector, int[] chunkLengths) {
+    public static long sumChunksLength(int[] chunkIds, Vec vec, Optional<Vec> weightsVector, int[] chunkLengths) {
+        assert chunkLengths.length == chunkIds.length;
+        
         for (int i = 0; i < chunkIds.length; i++) {
             final int chunk = chunkIds[i];
-            chunkLengths[i] = vec.chunkLen(chunk);
-            if (weightsVector == null)
-                continue;
-
-            Chunk weightVecChunk = weightsVector.chunkForChunkIdx(chunk);
-            if (weightVecChunk.atd(0) == 0) chunkLengths[i]--;
-            int nzIndex = 0;
-            do {
-                nzIndex = weightVecChunk.nextNZ(nzIndex, true);
-                if (nzIndex < 0 || nzIndex >= weightVecChunk._len) break;
-                if (weightVecChunk.atd(nzIndex) == 0) chunkLengths[i]--;
-            } while (true);
+            if (weightsVector.isPresent()) {
+                final Chunk weightVecChunk = weightsVector.get().chunkForChunkIdx(chunk);
```

Comment:
Can there be empty chunks?

Reply:
Great catch. What do you mean by empty chunks ? If you mean both empty, than this will work: I've added a test to demonstrate it. https://github.com/h2oai/h2o-3/pull/5038/commits/7de743c10ab986f076129385854cde456692f281

It is assumed both vectors have the same layout and thus both chunks are of the same size (0).

Reply:
The test works but only coincidentally. Empty chunks are typically represented by a constant chunk (eg C0LChunk) - these don't have any range checks - just return a constant, this approach thus has 2 issues:

- constant empty chunk can be represented by any constant - it is only a coincidence it is 0,
- there can be cases where empty chunks will not be represented by a constant chunk.

I will push a test that shows the issue.

Reply:
Added a trivial fix + added one more variant of the test.

My schedule is full of meetings now, but I"ll try to make this process parallel as more refined and push that variant as well (we can always go back).

Reply:
I think this fix is fine - no need for parallelization. This is not where we can gain much time.